File: c/algorithm/sort/quick_sort.c
/** Quick Sort
 *
 * O(NlogN), Omega(NlogN)
 * worst: O(n^2)
 *  e.g. sorted array
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

void quick_sort(int32_t arr[], uint32_t size);
uint32_t partition(int32_t arr[], int32_t left, int32_t right);
void sort(int32_t arr[], int32_t left, int32_t right);
void swap(int32_t arr[], int32_t i, int32_t j);

int32_t main(void)
{
    int32_t arr[] = { 5, 1, 2, 10, 8, 0, 23, 7 };
    uint32_t size = sizeof(arr) / sizeof(int32_t);

    quick_sort(arr, size);

    printf("{ ");
    for (uint32_t i = 0; i < size; i++)
    {
        printf("%d ", arr[i]);
    }
    printf("}\n");
}

/*
 * Function: quick_sort
 * -------------------
 * Execute sort function
 *
 *  params: int32_t arr[], uint32_t size
 *  returns: void
 */
void quick_sort(int32_t arr[], uint32_t size)
{
    sort(arr, 0, size);
}

/*
 * Function: sort
 * -------------------
 * Sort arr[] between left to right
 *
 *  params: int32_t arr[], int32_t left, int32_t right
 *  returns: void
 */
void sort(int32_t arr[], int32_t left, int32_t right)
{
    // base case
    if (left >= right)
    {
        return;
    }

    uint32_t pivot_idx = partition(arr, left, right);
    sort(arr, left, pivot_idx - 1);
    sort(arr, pivot_idx + 1, right);
}

/*
 * Function: partition
 * -------------------
 * Return pivot index
 *
 *  params: int32_t arr[], int32_t left, int32_t right
 *  returns: uint32_t
 */
uint32_t partition(int32_t arr[], int32_t left, int32_t right)
{
    int32_t mid = (left + right) >> 1;
    swap(arr, left, mid);

    int32_t pivot = arr[left];
    int32_t i = left, j = right;

    while (i < j)
    {
        while (pivot < arr[j])
        {
            j--;
        }
        while (pivot >= arr[i] && i < j)
        {
            i++;
        }
        swap(arr, i, j);
    }
    arr[left] = arr[i];
    arr[i] = pivot;

    return i;
}

/*
 * Function: swap
 * -------------------
 * Swap arr[] elements
 *
 *  params: int32_t arr[], int32_t a, int32_t b
 *  returns: void
 */
void swap(int32_t arr[], int32_t a, int32_t b)
{
    int32_t tmp = arr[a];
    arr[a] = arr[b];
    arr[b] = tmp;
}

File: java/algorithm/sort/QuickSort.java
package algorithm.sort;

/**
 * QuickSort
 * 
 * 평균 O(NlogN)
 */
public class QuickSort {

    public static void quickSort(int[] arr) {
        sort(arr, 0, arr.length-1);
    }

    private static void sort(int[] arr, int left, int right) {
        // base case
        if (left >= right) {
            return;
        }

        int pIdx = partition(arr, left, right);
        sort(arr, left, pIdx-1);
        sort(arr, pIdx+1, right);
    }

    private static int partition(int[] arr, int left, int right) {
        int mid = (left + right) >> 1;
        swap(arr, left, mid);

        int pivot = arr[left];
        int i = left, j = right;

        while (i < j) {
            while (pivot < arr[j]) {
                j--;
            }
            while (pivot >= arr[i] && i < j) {
                i++;
            }
            swap(arr, i, j);
        }
        arr[left] = arr[i];
        arr[i] = pivot;

        return i;
    }

    private static void swap(int[] arr, int a, int b) {
        int tmp = arr[a];
        arr[a] = arr[b];
        arr[b]= tmp;
    }

    public static void main(String[] args) {
        /**
         * 첫 partition() 실행 과정
         * arr = {80, 70, 60, 30, 10, 20, 40};
         * pivot = 30 = arr[(left + right) >> 1]
         * 
         * i                  j   : i = left, j = right
         * 30 70 60 80 10 20 40   : init
         *    i            j      : while 종료 시점
         * 30 20 60 80 10 70 40   : swap
         *       i      j         : while 종료 시점
         * 30 20 10 80 60 70 40   : swap
         *       ij               : while 종료 시점
         * 30 20 10 80 60 70 40   : swap
         *  --while문 종료--
         *       
         * 10 20 30 80 60 70 40   : swap arr[left] && arr[i]
         * 
         * return i = 2
        */
        int[] test = {80, 70, 60, 30, 10, 20, 40};
        quickSort(test);

        for (int v: test) {                     // prints 20 30 40 50 60 70 80
            System.out.print(v + " ");
        }
        System.out.println();
    }
}
